core/chains/evm/headtracker/head_saver.go

```diff
@@ -36,23 +38,26 @@ func (hs *headSaver) Save(ctx context.Context, head *evmtypes.Head) error {
 		return err
 	}
 
-	historyDepth := uint(hs.htConfig.HistoryDepth())
-	hs.heads.AddHeads(historyDepth, head)
+	hs.heads.AddHeads(head)
 
-	return hs.orm.TrimOldHeads(ctx, historyDepth)
+	return nil
 }
 
-func (hs *headSaver) Load(ctx context.Context) (chain *evmtypes.Head, err error) {
-	historyDepth := uint(hs.htConfig.HistoryDepth())
-	heads, err := hs.orm.LatestHeads(ctx, historyDepth)
+func (hs *headSaver) Load(ctx context.Context, latestFinalized int64) (chain *evmtypes.Head, err error) {
+	minBlockNumber := hs.calculateMinBlockToKeep(latestFinalized)
+	heads, err := hs.orm.LatestHeads(ctx, minBlockNumber)
 	if err != nil {
 		return nil, err
 	}
 
-	hs.heads.AddHeads(historyDepth, heads...)
+	hs.heads.AddHeads(heads...)
 	return hs.heads.LatestHead(), nil
 }
 
+func (hs *headSaver) calculateMinBlockToKeep(latestFinalized int64) int64 {
+	return max(latestFinalized-int64(hs.htConfig.HistoryDepth()), 0)
```

Comment:
Is it worth using HistoryDepth in two different ways to not store additional 50 blocks (or 1500 in case of Polygon)?
It's kind of confusing that with finality tag enabled we keep `history depth` of finalized blocks, but when it's not enabled we keep only `history depth-finalityDepth` of finalized blocks

Reply:
The problem I see is that for finalityDepth, it is easy to ensure HistoryDepth is always larger.
But for FinalityTags, it is pretty hard to know the right HistoryDepth which will always be more than the finalized block, if we just use the HistoryDepth as is.
What if we set HistoryDepth for a chain to 100, but the FinalityTags suddenly give us latest finalized head to be 150 deep?

Reply:
We don't have to ensure anything. `HistoryDepth` is going to work differently in the new Head Tracker. Head Tracker will **always** store up to the latest finalized block whether it uses finality tag or finality depth. From there anyone who wants additional history can set the `HistoryDepth` value to fetch it. This way there is no problem with how long the finalized block is. I expect most users to set the `HistoryDepth` to 0 since they don't care about additional heads. We can simply update the configs to reflect this change.

Reply:
@dimriou This is not how HistoryDepth is used today.
Currently we enforce through config validation that HistoryDepth > FinalityDepth.
What you are suggesting makes sense, but will require us to change the meaning and default values of HistoryDepth on all chains.

Reply:
Yeah that's what I'm suggesting. `HistoryDepth` isn't client facing, so users don't really know what to do with it. They only care about Finality, so as long as we can keep that we should change `HistoryDepth` to something more meaningful. And we can change the validations from here: https://github.com/smartcontractkit/chainlink/blob/develop/core/chains/evm/config/toml/config.go#L390